modules/dnstap/dnstap.c

```diff
@@ -0,0 +1,410 @@
+ * GNU General Public License for more details.
+
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * @file dnstap.c
+ * @brief dnstap based query logging support
+ *
+ * dnstap module currently supports logging dns responses to a unix socket
+ * in dnstap format using fstrm framing library. The unix socket and the
+ * socket reader should be present before starting kresd
+ *
+ * Tunables:
+ * sockpath: is the the unix socket file where dnstap messages will be send
+ * logRespPkt: if true responses in wire format will be logged
+ *
+ * sample configuration for dnstap
+ * dnstap = {
+ *     sockPath = "/tmp/dnstap.sock",
+ *     logRespPkt = true
+ * }
+ *
+ */
+
+#include "lib/module.h"
+#include "lib/layer.h"
+#include "lib/resolve.h"
+#include "contrib/dnstap/dnstap.pb-c.h"
+#include <ccan/json/json.h>
+#include <fstrm.h>
+
+#define DEBUG_MSG(fmt, ...) kr_log_verbose("[dnstap] " fmt, ##__VA_ARGS__);
+#define CFG_SOCK_PATH "sockPath"
+#define CFG_LOG_RESP_PKT "logRespPkt"
+#define DEFAULT_SOCK_PATH "/tmp/dnstap.sock"
+#define DNSTAP_CONTENT_TYPE "protobuf:dnstap.Dnstap"
+#define DNSTAP_INITIAL_BUF_SIZE         256
+
+/* Internal data structure */
+struct dnstap_data {
+	bool log_resp_pkt;
+	struct fstrm_iothr *iothread;
+	struct fstrm_iothr_queue *ioq;
+};
+
+/*
+ * dt_pack packs the dnstap message for transport
+ * https://gitlab.labs.nic.cz/labs/knot/blob/master/src/contrib/dnstap/dnstap.c#L24
+ * */
+uint8_t* dt_pack(const Dnstap__Dnstap *d, uint8_t **buf, size_t *sz)
+{
+	ProtobufCBufferSimple sbuf = { { NULL } };
+
+	sbuf.base.append = protobuf_c_buffer_simple_append;
+	sbuf.len = 0;
+	sbuf.alloced = DNSTAP_INITIAL_BUF_SIZE;
+	sbuf.data = malloc(sbuf.alloced);
+	if (sbuf.data == NULL) {
+		return NULL;
+	}
+	sbuf.must_free_data = true;
+
+	*sz = dnstap__dnstap__pack_to_buffer(d, (ProtobufCBuffer *) &sbuf);
+	*buf = sbuf.data;
+	return *buf;
+}
+
+/* set_address fills in address deatail in dnstap_message 
+ * https://gitlab.labs.nic.cz/labs/knot/blob/master/src/contrib/dnstap/message.c#L28
+ */
+static void set_address(const struct sockaddr *sockaddr,
+		ProtobufCBinaryData   *addr,
+		protobuf_c_boolean    *has_addr,
+		uint32_t              *port,
+		protobuf_c_boolean    *has_port) {
+	if (sockaddr == NULL) {
+		*has_addr = 0;
+		*has_port = 0;
+		return;
+	}
+
+	*has_addr = true;
+	*has_port = true;
+
+	if (sockaddr->sa_family == AF_INET) {
+		const struct sockaddr_in *sai;
+		sai = (const struct sockaddr_in *)sockaddr;
+		addr->len = sizeof(sai->sin_addr);
+		addr->data = (uint8_t *)&sai->sin_addr.s_addr;
+		*port = ntohs(sai->sin_port);
+	} else if (sockaddr->sa_family == AF_INET6) {
+		const struct sockaddr_in6 *sai6;
+		sai6 = (const struct sockaddr_in6 *)sockaddr;
+		addr->len = sizeof(sai6->sin6_addr);
+		addr->data = (uint8_t *)&sai6->sin6_addr.s6_addr;
+		*port = ntohs(sai6->sin6_port);
+	}
```

Comment:
There are convenience functions for this, please reuse:
* https://github.com/CZ-NIC/knot-resolver/blob/master/lib/utils.c#L245-L270
* https://github.com/CZ-NIC/knot-resolver/blob/master/lib/utils.h#L129-L137
(and add similar one for port)